panoply/datasource.py

```diff
@@ -1,24 +1,67 @@
+from abc import abstractmethod, ABCMeta
 import base64
 import traceback
 from concurrent.futures import ThreadPoolExecutor
 from functools import wraps
 from threading import Event
+from typing import List
 
 import backoff
 import requests
 
 from . import events
-from .errors.exceptions import TokenValidationException
+from .errors.exceptions import (TokenValidationException, UnableToListPanoplyResources, UnableToGetPanoplyResource)
+from .records import PanoplyRecordGroup
+from .resources import PanoplyResource
 
 
-class DataSource(events.Emitter):
+class DataSource(events.Emitter, metaclass=ABCMeta):
     """ A base DataSource object """
 
-    def __init__(self, source, options={}):
+    def __init__(self, source, options={}, validate_methods=True):
         super(DataSource, self).__init__()
 
         self.source = source
         self.options = options
+        if validate_methods:
+            self.validate_metadata_methods()
+
+    @abstractmethod
+    def read(self, batch_size=None) -> List[PanoplyRecordGroup]:
+        """
+        Reads data from the sources and returns it as records
+        """
+        return
```

Comment:
```suggestion
    @abstractmethod
    def read(self, batch_size=None) -> List[PanoplyRecordGroup]:
        """
        Reads data from the sources and returns it as records
        """
```

You can also remove `return` in empty method that has docstring